File: go/piece/piece.go
package piece

import (
	"math/rand"

	"github.com/ni3po42/LearnByTetris/go/board"
	"github.com/ni3po42/LearnByTetris/go/common"
)

type PieceId byte

type pieceData struct {
	field    [16]byte
	width    int
	height   int
	ordinals byte
	color    common.CellData
}

var rawPieces = map[PieceId]pieceData{
	'T': {
		field: [16]byte{
			0, 1, 0,
			1, 1, 1,
			0, 0, 0,
		},
		width:    3,
		height:   3,
		ordinals: 4,
		color:    1,
	},
	'O': {
		field: [16]byte{
			1, 1,
			1, 1,
		},
		width:    2,
		height:   2,
		ordinals: 1,
		color:    2,
	},
	'S': {
		field: [16]byte{
			0, 1, 1,
			1, 1, 0,
			0, 0, 0,
		},
		width:    3,
		height:   3,
		ordinals: 2,
		color:    3,
	},
	'Z': {
		field: [16]byte{
			1, 1, 0,
			0, 1, 1,
			0, 0, 0,
		},
		width:    3,
		height:   3,
		ordinals: 2,
		color:    4,
	},
	'J': {
		field: [16]byte{
			0, 0, 0,
			1, 1, 1,
			0, 0, 1,
		},
		width:    3,
		height:   3,
		ordinals: 4,
		color:    5,
	},
	'L': {
		field: [16]byte{
			0, 0, 0,
			1, 1, 1,
			1, 0, 0,
		},
		width:    3,
		height:   3,
		ordinals: 4,
		color:    6,
	},
	'I': {
		field: [16]byte{
			0, 0, 0, 0,
			1, 1, 1, 1,
			0, 0, 0, 0,
			0, 0, 0, 0,
		},
		width:    4,
		height:   4,
		ordinals: 2,
		color:    7,
	},
}

const idLookup = "TOSZJLI"

type Piece struct {
	row, col         int
	prevRow, prevCol int
	turns            byte
	prevTurns        byte
	active           bool

	height, width int
	id            PieceId
	ordinals      byte
	color         common.CellData
}

func NewPiece(args ...PieceId) *Piece {
	piece := Piece{
		id:        ' ',
		turns:     0,
		active:    false,
		row:       0,
		col:       0,
		prevCol:   0,
		prevRow:   0,
		prevTurns: 0,
		width:     0,
		height:    0,
		ordinals:  1,
		color:     0,
	}

	if len(args) > 0 {
		piece.id = args[0]
		piece.width = rawPieces[piece.id].width
		piece.height = rawPieces[piece.id].height
		piece.ordinals = rawPieces[piece.id].ordinals
		piece.color = rawPieces[piece.id].color
		piece.active = true
	}

	return &piece
}

func GetRandomPieceId() PieceId {
	index := rand.Intn(7)
	return PieceId(idLookup[index])
}

func Pieces() func() (*Piece, PieceId) {

	currentId := GetRandomPieceId()

	return func() (*Piece, PieceId) {
		nextId := GetRandomPieceId()
		currentPiece := NewPiece(currentId)
		currentId = nextId
		return currentPiece.Play(), nextId
	}
}

func (p *Piece) save() {
	p.prevRow = p.row
	p.prevCol = p.col
	p.prevTurns = p.turns
}

func (p Piece) Position() (int, int) {
	return p.row, p.col
}

func (p Piece) Dimensions() (int, int) {
	if p.turns%2 == 0 {
		return p.height, p.width
	} else {
		return p.width, p.height
	}
}

func (p Piece) GetID() PieceId {
	return p.id
}

func (p *Piece) Revert() {
	p.row = p.prevRow
	p.col = p.prevCol
	p.turns = p.prevTurns
}

func (p *Piece) Play() *Piece {
	p.row = 0
	p.col = (board.COLS / 2) - 1
	p.prevRow = 0
	p.prevCol = p.col
	p.turns = 0
	return p
}

func (p *Piece) Rotate(direction common.Direction) {
	p.save()

	switch direction {
	case common.LEFT:
		p.turns--
		p.turns = (p.turns + p.ordinals) % p.ordinals
	case common.RIGHT:
		p.turns++
		p.turns = p.turns % p.ordinals
	}
}

func (p *Piece) Drop() {
	p.save()
	p.row++
}

func (p *Piece) Move(direction common.Direction) {
	p.save()

	switch direction {
	case common.LEFT:
		p.col--
	case common.RIGHT:
		p.col++
	}
}

func (p *Piece) Commit() {
	p.active = false
}

func (p Piece) Scan(useNeturalPosition bool) func() (bool, common.ICellInfo) {

	row, col := p.row, p.col
	height, width := p.height, p.width
	turns := p.turns
	bound := p.width * p.height
	rawPieceData := rawPieces[p.id]
	active := p.active
	color := p.color

	var index int = 0

	return func() (bool, common.ICellInfo) {
		var r, c int
		var cellData common.CellData

		for {

			if index == bound {
				return false, nil
			}

			if rawPieceData.field[index] == 0 {
				index++
			} else {
				break
			}
		}

		if active {
			cellData = color | board.ACTIVE_MASK
		} else {
			cellData = color | board.STATIC_MASK
		}

		switch turns {
		case 0:
			r, c = index/width, index%width
		case 1:
			r, c = index%width, height-(index/width)-1
		case 2:
			r, c = (bound-index-1)/width, (bound-index-1)%width
		case 3:
			r, c = width-(index%width)-1, index/width
		}

		index++
		if useNeturalPosition {
			return true, board.NewScanData(r, c, cellData)
		} else {
			return true, board.NewScanData(r+row, c+col, cellData)
		}
	}
}

File: c/piece/piece.c
#include <stdbool.h>
#include <stdio.h>

#include "../generator/generator.h"
#include "../constants.h"

#include "../utilities.h"

#include "piece.h"
#include "piece.internal.h"

/**
 * private apis
*/

// defines data for each piece
RawPieceData RawPieces[] = {
    {
        .field = {
            0,1,0,
            1,1,1,
            0,0,0
        },
        .height = 3,
        .width = 3,
        .ordinals = 4,
        .id = PIECE_T,
        .color = 1
    },
    {
        .field = {
            1,1,
            1,1
        },
        .height = 2,
        .width = 2,
        .ordinals = 1,
        .id = PIECE_O,
        .color = 2
    },
    {
        .field = {
            0,1,1,
            1,1,0,
            0,0,0
        },
        .height = 3,
        .width = 3,
        .ordinals = 2,
        .id = PIECE_S,
        .color = 3
    },
    {
        .field = {
            1,1,0,
            0,1,1,
            0,0,0
        },
        .height = 3,
        .width = 3,
        .ordinals = 2,
        .id = PIECE_Z,
        .color = 4
    },
    {
        .field = {
            0,0,0,
            1,1,1,
            0,0,1
        },
        .height = 3,
        .width = 3,
        .ordinals = 4,
        .id = PIECE_J,
        .color = 5
    },
    {
        .field = {
            0,0,0,
            1,1,1,
            1,0,0
        },
        .height = 3,
        .width = 3,
        .ordinals = 4,
        .id = PIECE_L,
        .color = 6
    },
    {
        .field = {
            0,0,0,0,
            1,1,1,1,
            0,0,0,0,
            0,0,0,0
        },
        .height = 4,
        .width = 4,
        .ordinals = 2,
        .id = PIECE_I,
        .color = 7
    }
};

// initializes a piece to be used on the game board
// static void initPiece(Piece* piece) {
   
// }

// builds piece based on raw piece data
static void constructPiece(Piece* piece, RawPieceData* data) {
    if (piece == NULL){
        return;
    }
    
    piece->data = data->field;
    piece->scanWidth = data->width;
    piece->scanHeight = data->height;
    piece->id = data->id;
    piece->color = data->color;
    
    piece->_ordinals = data->ordinals;
    piece->_direction = 0;
    piece->_active = true;
    
    piece->_row = 0;
    piece->_col = 0;
    piece->_prevRow = 0;
    piece->_prevCol = 0;
    piece->_prevDirection = 0;
}

// save state of piece
static void savePieceState(Piece* piece) {
    if (piece == NULL){
        return;
    }
    piece->_prevRow = piece->_row;
    piece->_prevCol = piece->_col;
    piece->_prevDirection = piece->_direction;    
}

// copies state of one piece to another
static void promotePiece(Piece* source, Piece* destination) {
    destination->color = source->color;
    destination->data = source->data;
    destination->id = source->id;
    destination->scanHeight = source->scanHeight;
    destination->scanWidth = source->scanWidth;
    destination->_active = source->_active;
    //destination->_col = source->_col;
    //destination->_row = source->_row;
    destination->_direction = source->_direction;
    destination->_ordinals = source->_ordinals;
    //destination->_prevCol = source->_prevCol;
    //destination->_prevDirection = source->_prevDirection;
    //destination->_prevRow = source->_prevRow;
    
    destination->_row = 0;
    destination->_col = (BOARD_COLS / 2) - 1;
    destination->_prevRow = 0;
    destination->_prevCol = (BOARD_COLS / 2) - 1;
    destination->_prevDirection = 0;
}

// generator function for creating infinite sequence of pieces
static void nextPiece_genFunc(GeneratorHandle* handle) {
    gen_restore(handle, Piece, currentPiece);
    gen_restore(handle, Piece, nextPiece);       
    gen_frame_init(handle, void, argument);
    
    int nextIndex = getRandomIntFromRange(0,6);              
    constructPiece(nextPiece, &RawPieces[nextIndex]);

    while(true) {
        promotePiece(nextPiece, currentPiece);
        // initPiece(currentPiece);
        
        nextIndex = getRandomIntFromRange(0,6);              
        constructPiece(nextPiece, &RawPieces[nextIndex]);
        // to tomorrow tim, since we are copying nextPieceData literally,
        // it means the pointers are copied, we think that by the time we
        // get the value, the data the pointers are pointing to have already
        // changed!
        NextPieceData nextPieceData = {
            .currentPiece = currentPiece,
            .nextPiece = nextPiece
        };
            
        gen_yield(handle, &nextPieceData);
    }    
}

/**
 * public apis
*/

void revertPiece(Piece* piece) {
    if (piece == NULL){
        return;
    }
    piece->_row = piece->_prevRow;
    piece->_col = piece->_prevCol;
    piece->_direction = piece->_prevDirection;    
}

void rotatePieceLeft(Piece* piece) {
    if (piece == NULL){
        return;
    }
    savePieceState(piece);
    piece->_direction = (--(piece->_direction) + piece->_ordinals) % piece->_ordinals;    
}

void rotatePieceRight(Piece* piece) {
    if (piece == NULL){
        return;
    }
    savePieceState(piece);
    piece->_direction = (++(piece->_direction)) % piece->_ordinals;
}

void dropPiece(Piece* piece) {
    if (piece == NULL){
        return;
    }
    savePieceState(piece);
    piece->_row++;
}

void leftPiece(Piece* piece) {
    if (piece == NULL){
        return;
    }
    savePieceState(piece);
    piece->_col -= 1;    
}

void rightPiece(Piece* piece) {
    if (piece == NULL){
        return;
    }
    savePieceState(piece);
    piece->_col += 1;     
}

void commitPiece(Piece* piece) {
    if (piece == NULL){
        return;
    }
    piece->_active = false;
}

int getPieceWidth(const Piece* piece) {
    if (piece == NULL){
        return -1;
    }
    return piece->_direction % 2 ? piece->scanHeight : piece->scanWidth;    
}

int getPieceHeight(const Piece* piece) {
    if (piece == NULL){
        return -1;
    }
    return piece->_direction % 2 ? piece->scanWidth : piece->scanHeight;
}

int getPieceRow(const Piece* piece) {
    if (piece == NULL){
        return -1;
    }
    return piece->_row;
}

int getPieceCol(const Piece* piece) {
    if (piece == NULL){
        return -1;
    }
    return piece->_col;
}


int getPieceId(const Piece* piece) {
    if (piece == NULL) {
        return -1;
    }
    return piece->id;
}

color_t getPieceColor(const Piece* piece) {
    if (piece == NULL) {
        return -1;
    }
    return piece->color;
}

void pieceScan(const Piece* piece, bool posNeutral, PieceScanFunc func) {
    
    int bound = piece->scanWidth * piece->scanHeight;
    int scanWidth = piece->scanWidth;
    int index;
    int r, c;
    color_t color;
    
    for(index = 0; index < bound; index++) {
        color = (piece->data)[index] ? piece->color : 0x00;
        
        if (color == 0){
            continue;
        }

        color |= piece->_active ? ACTIVE_MASK : STATIC_MASK;

        // based on piece's direction, actual row/col values can be different
        // this switch figures out when row/col values are swapped and account
        // for size of bounding box
        switch(piece->_direction) {
            case 0:
                r = index / scanWidth;
                c = index % scanWidth;
                break;
            case 1:
                r = index % scanWidth;
                c = piece->scanHeight - (index / scanWidth) - 1;
                break;
            case 2:
                r = (bound - index - 1) / scanWidth;
                c = (bound - index - 1) % scanWidth;                    
                break;
            case 3:
                r = scanWidth - (index % scanWidth) - 1;
                c = index / scanWidth;                    
                break;
        }

        func(
            r + (posNeutral ? 0 :piece->_row),
            c + (posNeutral ? 0 :piece->_col),
            color
        );
    }
    
}

GeneratorHandle* pieces() {
    return gen_func(nextPiece_genFunc, NextPieceData, NULL);
}